### PR TITLE
Adds documentation for editor.js

### LINKED ISSUE #1: Document wp-admin/js/editor.js

### PR DESCRIPTION
**Pull request by terw-dan**
Replaces #22 

This is a new PR to make sure we don't have conflicts with master. 

Give props to @IreneStr as well. 

Fixes #1

### CODE REVIEW

File: src/wp-admin/js/editor.js [hunk 1]
```diff
@@ -224,102 +346,221 @@
...
+			/*
+			 * Replaces a newline followed by one or multiple spaces
+			 * and another newline with 2 newlines.
+			 */
 			text = text.replace( /\n\s*\n+/g, '\n\n' );
+
+			// Replaces 2 newlines with a paragraph and a single newline.
 			text = text.replace( /([\s\S]+?)\n\n/g, '<p>$1</p>\n' );
+
+			// Replaces p tags containing only white spaces with nothing.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Remove empty paragraphs.

File: src/wp-admin/js/editor.js [hunk 2]
```diff
@@ -224,102 +346,221 @@
...
 			text = text.replace( /<p>\s*?<\/p>/gi, '');
+
+			// Removes spaces and p tags around block level elements.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: p tags => `<p>` elements

File: src/wp-admin/js/editor.js [hunk 3]
```diff
@@ -224,102 +346,221 @@
...
 			text = text.replace( new RegExp( '<p>\\s*(</?(?:' + blocklist + ')(?: [^>]*)?>)\\s*</p>', 'gi' ), '$1' );
+
+			// Removes p tags around li elements.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: ^ see above

File: src/wp-admin/js/editor.js [hunk 4]
```diff
@@ -224,102 +346,221 @@
...
 			text = text.replace( /<p>(<li.+?)<\/p>/gi, '$1');
+
+			// Removes spaces and p tags from blockquotes.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: ^ see above

Comment (moorscode) [inline src/wp-admin/js/editor.js]: Places the `<blockquote>` outside of the paragraph.

File: src/wp-admin/js/editor.js [hunk 5]
```diff
@@ -224,102 +346,221 @@
...
 			text = text.replace( /<p>\s*<blockquote([^>]*)>/gi, '<blockquote$1><p>');
+
+			// Removes spaces and switches the p tag and blockquote.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: ^ see above

File: src/wp-admin/js/editor.js [hunk 6]
```diff
@@ -224,102 +346,221 @@
...
 			text = text.replace( /<\/blockquote>\s*<\/p>/gi, '</p></blockquote>');
+
+			// Removes spaces at the start and p tags from block level elements.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: `<p>` elements

File: src/wp-admin/js/editor.js [hunk 7]
```diff
@@ -224,102 +346,221 @@
...
 			text = text.replace( new RegExp( '<p>\\s*(</?(?:' + blocklist + ')(?: [^>]*)?>)', 'gi' ), '$1' );
+
+			// Removes spaces at the end and p tags from block level elements.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: ^ see above

File: src/wp-admin/js/editor.js [hunk 8]
```diff
@@ -224,102 +346,221 @@
...
 			text = text.replace( new RegExp( '(</?(?:' + blocklist + ')(?: [^>]*)?>)\\s*</p>', 'gi' ), '$1' );
 
-			// Remove redundant spaces and line breaks after existing <br /> tags
+			// Removes spaces and newlines after a br tag.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: ^ see above

File: src/wp-admin/js/editor.js [hunk 9]
```diff
@@ -224,102 +346,221 @@
...
 			text = text.replace( /(<br[^>]*>)\s*\n/gi, '$1' );
 
-			// Create <br /> from the remaining line breaks
+			// Replaces spaces followed by a newline with a br tag followed by a new line.
 			text = text.replace( /\s*\n/g, '<br />\n');
 
+			// Removes whitespaces and a single br tag after a block level element.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Remove `<br>` element that follows a block element directly, ignoring spaces.

File: src/wp-admin/js/editor.js [hunk 10]
```diff
@@ -224,102 +346,221 @@
...
+
+			// Removes spaces at the end and p tags from block level elements.
 			text = text.replace( new RegExp( '(</?(?:' + blocklist + ')(?: [^>]*)?>)\\s*</p>', 'gi' ), '$1' );
 
-			// Remove redundant spaces and line breaks after existing <br /> tags
+			// Removes spaces and newlines after a br tag.
 			text = text.replace( /(<br[^>]*>)\s*\n/gi, '$1' );
 
-			// Create <br /> from the remaining line breaks
+			// Replaces spaces followed by a newline with a br tag followed by a new line.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: ^ see above

File: src/wp-admin/js/editor.js [hunk 11]
```diff
@@ -224,102 +346,221 @@
...
 			text = text.replace( /\s*\n/g, '<br />\n');
 
+			// Removes whitespaces and a single br tag after a block level element.
 			text = text.replace( new RegExp( '(</?(?:' + blocklist + ')[^>]*>)\\s*<br />', 'gi' ), '$1' );
+
+			/*
+			 * Removes a br tag preceding white spaces followed by a
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Remove `<br>` element that is followed by one of ... elements, ignoring spaces.

File: src/wp-admin/js/editor.js [hunk 12]
```diff
@@ -224,102 +346,221 @@
...
+			 * p, li, div, dl, dd, dt, th, pre, td, ul, or ol element.
+			 */
 			text = text.replace( /<br \/>(\s*<\/?(?:p|li|div|dl|dd|dt|th|pre|td|ul|ol)>)/gi, '$1' );
+
+			// Removes white spaces, p tags and br tags in captions.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: `<p>` and `<br>` elements

File: src/wp-admin/js/editor.js [hunk 13]
```diff
@@ -224,102 +346,221 @@
...
+			 * @param {string} a The complete match.
+			 * @param {string} b The first capture group.
+			 * @param {string} c The second capture group.
+			 * @returns {string} The string in paragraph tags.
+			 */
 			text = text.replace( /(<(?:div|th|td|form|fieldset|dd)[^>]*>)(.*?)<\/p>/g, function( a, b, c ) {
+
+				/*
+				 * Checks if the matched group has a p open tag in it, if so we don't need to add
+				 * another one and the complete match can be returned.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Suggestion: so we don't need to enclose it with a paragraph.

File: src/wp-admin/js/editor.js [hunk 14]
```diff
@@ -224,102 +346,221 @@
...
 
+				/*
+				 * If there is no p open tag in the matched string,
+				 * add it and return the string including p tags.
+				 */
 				return b + '<p>' + c + '</p>';
 			});
 
-			// put back the line breaks in pre|script
+			// Puts back the line breaks in pre|script.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Restores the line breaks in `<pre>`, `<script>` elements

File: src/wp-admin/js/editor.js [hunk 15]
```diff
@@ -224,102 +346,221 @@
...
 			if ( preserve_linebreaks ) {
 				text = text.replace( /<wp-line-break>/g, '\n' );
 			}
 
+			// Puts back the br-tags in captions.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: ^ See above (restores)

File: src/wp-admin/js/editor.js [hunk 16]
```diff
@@ -224,102 +346,221 @@
...
 
 			return text;
 		}
 
-		// Add old events
+		/**
+		 * @summary Modifies the data when a switch is made from HTML to text.
+		 *
+		 * Modifies the data when a switch is made.
+		 * Runs removep to remove the p tags from text.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Just say: Remove the p tags from text.

Comment (moorscode) [inline src/wp-admin/js/editor.js]: The `removep` is an implementation detail

File: src/wp-admin/js/editor.js [hunk 17]
```diff
@@ -352,11 +606,16 @@
 			return obj.data;
 		}
 
+		// Runs the init function on document ready if jQuery is available.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Binds the init function to be run when the document is loaded.

File: src/wp-admin/js/editor.js [hunk 18]
```diff
@@ -352,11 +606,16 @@
...
 		if ( $ ) {
 			$( document ).ready( init );
 		} else if ( document.addEventListener ) {
+
+			// Uses the addEventListener to bind the init event on document load.
 			document.addEventListener( 'DOMContentLoaded', init, false );
 			window.addEventListener( 'load', init, false );
+
+			// Uses the addEvent to bind the init event on document load.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Would move this comment inside of the following `else`

File: src/wp-admin/js/editor.js [hunk 19]
```diff
@@ -382,5 +645,8 @@
 		return exports;
 	}
 
+	/**
+	 * @namespace {SwitchEditors} switchEditors
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Expose the switch editors to be used globally.

File: src/wp-admin/js/editor.js [hunk 20]
```diff
@@ -22,6 +53,15 @@
...
 		}
 
+		/**
+		 * @summary Retrieves the height of the toolbar.
+		 *
+		 * @since 3.7.2
+		 *
+		 * @param {Object} editor The tinyMCE editor.
+		 * @returns {number} If the height is between 10 and 200 return the height,
+		 * 								else return 30.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Get the height of the editor based on the container it is placed in.

File: src/wp-admin/js/editor.js [hunk 21]
```diff
@@ -33,6 +73,20 @@
 			return 30;
 		}
 
+		/**
+		 * @summary Switches the editor between visual and text
+		 *                     based on which button is pressed.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Strange indentation

Comment (moorscode) [inline src/wp-admin/js/editor.js]: This line can be removed totally

File: src/wp-admin/js/editor.js [hunk 22]
```diff
@@ -33,6 +73,20 @@
...
+		 *
+		 * @since 2.5
+		 *
+		 * @memberof switchEditors
+		 *
+		 * @param {string} id			The id of the editor you want to change the editor mode for.
+		 *											If an undefined id is given, it defaults to content.
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: `If no id is given` instead of `If an undefined id is given`

File: src/wp-admin/js/editor.js [hunk 23]
```diff
@@ -43,6 +97,10 @@
 				$textarea = $$( '#' + id ),
 				textarea = $textarea[0];
 
+			/*
+			 * If the mode is toggle, checks the current state of the editor
```
Comment (moorscode) [inline src/wp-admin/js/editor.js]: Toggle the mode between `visual` and `textual` representation.

File: src/wp-admin/js/editor.js (hunk 20 above)
Comment (terw-dan) [inline src/wp-admin/js/editor.js]: This specifically gets the height of the toolbar, not the complete editor.

**Pull request closed without merging**